### ISSUE TITLE
Rule formatting tool.

### ISSUE THREAD

**Issue by juliusv** [prometheus/prometheus#21]
Like "gofmt" for Go, we ought to have a "promfmt" for Prometheus since we have a syntax tree. The idea being that the system produces uniform style that minimizes deviation and learning curve.

**Update after we have totally moved to YAML rule files**: In addition to formatting the PromQL expressions, we also want to format the YAML files to have a fixed structure, while preserving comments for both PromQL expressions and the YAML file.

**Comment by grobie** [comment]
:+1: * :100:

**Comment by fabxc** [comment]
Yep, that'd be great. Preserving comments is the biggest (and basically only) problem.

**Comment by brian-brazil** [comment]
Aren't we in the middle of adding this to promtool?

**Comment by mattbostock** [comment]
Related: https://github.com/prometheus/prometheus/pull/1779

**Comment by beorn7** [comment]
Some observations:
- With the 2.x rule format (YAML embedding PromQL), we want this to act on YAML files, too. 
- It should preserve both comments in YAML and in PromQL.
- Indentation is crucial (not just line breaks).
- It has to be seen if certain line breaks in the input should be preserved in the output (cf. how `gofmt` does it).
- Wherever a rule is rendered, the same formatting should apply, in particular on the _Alerts_ and _Rules_ tabs of the Prometheus web UI. This implies that the AST needs to keep comments (and possibly certain line breaks, see previous point).

**Comment by brian-brazil** [comment]
> It should preserve both comments in YAML

Our current YAML library doesn't do this, but is hoping to "soon".

**Comment by geekodour** [comment]
I'd like to include this in my gsoc proposal. @brian-brazil, can you link to the YAML library you're referring to?

**Comment by brian-brazil** [comment]
https://github.com/go-yaml/yaml, but I'd scope this to just the PromQL as this stands.

**Comment by juliusv** [comment]
@brian-brazil But since the PromQL from rules is always embedded in YAML files, it at least has to preserve YAML comments, right? Otherwise the tool wouldn't be very useful as nobody wants to lose their comments the benefit of formatting.

**Comment by brian-brazil** [comment]
That'd be nice, but we can get benefits without that. I'd rather also not depend on something with an unclear timeline for a gsoc project.

**Comment by beorn7** [comment]
Let's start with the PromQL part. The YAML part should be almost trivial once the library supports comment nodes.

**Comment by juliusv** [comment]
Ok!

**Comment by brian-brazil** [comment]
v3 of the YAML library is out, which should allow for all of this.

**Comment by sylr** [comment]
FYI perfectly readable rule in configuration

```
(
    kube_job_status_failed > 0
    UNLESS kube_job_status_succeeded > 0
)
* on (namespace, job_name) group_left(maintainer) label_replace(kube_job_labels, "maintainer", "$1", "label_maintainer", "(.*)")
* on (namespace, job_name) group_left(pager) label_replace(kube_job_labels, "pager", "$1", "label_pager", "(.*)")
* on (namespace, job_name) group_left(paging) label_replace(kube_job_labels, "paging", "$1", "label_paging", "(.*)")

```

VS its garbage representation in the Web UI

```
(kube_deployment_status_replicas_available
  / kube_deployment_spec_replicas * 100) * on(namespace, deployment) group_left(maintainer)
  label_replace(kube_deployment_labels, "maintainer", "$1", "label_maintainer",
  "(.*)") * on(namespace, deployment) group_left(pager) label_replace(kube_deployment_labels,
  "pager", "$1", "label_pager", "(.*)") * on(namespace,
  deployment) group_left(paging) label_replace(kube_deployment_labels, "paging",
  "$1", "label_paging", "(.*)") < 100
```

**Comment by geekodour** [comment]
@sylr I think you posted some other rule from the web UI. also @beorn7 I am starting to work on this issue.

here's what it shows for me for your rule:
```
(kube_job_status_failed
  > 0 unless kube_job_status_succeeded > 0) * on(namespace, job_name) group_left(maintainer)
  label_replace(kube_job_labels, "maintainer", "$1", "label_maintainer",
  "(.*)") * on(namespace, job_name) group_left(pager) label_replace(kube_job_labels,
  "pager", "$1", "label_pager", "(.*)") * on(namespace,
  job_name) group_left(paging) label_replace(kube_job_labels, "paging", "$1",
  "label_paging", "(.*)")
```

**Comment by sylr** [comment]
@geekodour Yes, I mixed up two rules.

**Comment by haibeey** [comment]
since this is still open and added in project ideas for upcoming GSOC (https://github.com/cncf/soc#rule-formatting-tool) i would like to work on the completion of this during GSOC.  Any recent update and additional info regarding this would be appreciated @codesome  @brian-brazil @juliusv @geekodour @simonpasquier @sylr .
From reading the comment and other reference link it seems the next action point is building a formatter for promQL expression with the promQL AST?

**Comment by slrtbtfs** [comment]
@haibeey There already has been some progress in that direction.

When building such a formatter it probably would be best to start with what is already implemented [here](https://github.com/prometheus/prometheus/blob/master/promql/printer.go) and add proper white space and comment handling.

It would be also nice, when such features could be integrated with the upcoming [PromQL language server](https://github.com/prometheus-community/promql-langserver). Inside the language server code a lot of the stuff that is needed to format YAML files is already implemented, and it might make sense to reuse some of it.

**Comment by haibeey** [comment]
alright! Thanks @slrtbtfs . i will start perusing the links shared ASAP .

**Comment by Harkishen-Singh** [comment]
Can I take up this issue? Would love to implement this.

**Comment by roidelapluie** [comment]
It looks like @haibeey is already somehow planning working on this.

**Comment by Harkishen-Singh** [comment]
@roidelapluie I have already implemented a part of it way before. It's just that I had not mentioned here.

**Comment by haibeey** [comment]
> FYI perfectly readable rule in configuration
> 
> ```
> (
>     kube_job_status_failed > 0
>     UNLESS kube_job_status_succeeded > 0
> )
> * on (namespace, job_name) group_left(maintainer) label_replace(kube_job_labels, "maintainer", "$1", "label_maintainer", "(.*)")
> * on (namespace, job_name) group_left(pager) label_replace(kube_job_labels, "pager", "$1", "label_pager", "(.*)")
> * on (namespace, job_name) group_left(paging) label_replace(kube_job_labels, "paging", "$1", "label_paging", "(.*)")
> ```
> 
> VS its garbage representation in the Web UI
> 
> ```
> (kube_deployment_status_replicas_available
>   / kube_deployment_spec_replicas * 100) * on(namespace, deployment) group_left(maintainer)
>   label_replace(kube_deployment_labels, "maintainer", "$1", "label_maintainer",
>   "(.*)") * on(namespace, deployment) group_left(pager) label_replace(kube_deployment_labels,
>   "pager", "$1", "label_pager", "(.*)") * on(namespace,
>   deployment) group_left(paging) label_replace(kube_deployment_labels, "paging",
>   "$1", "label_paging", "(.*)") < 100
> ```

Is this formatting style acceptable? @codesome @juliusv @brian-brazil @beorn7 
I'm writing my proposal based on this style.

**Comment by codesome** [comment]
The final representation after formatting can be discussed and decided during the GSoC period (maybe even now if any maintainer wants to chime in, but I don't have the bandwidth to review it at the moment). I would suggest (to all GSoC aspirants) to focus on _how_ would you achieve this.

**Comment by brian-brazil** [comment]
https://prometheus.io/docs/practices/rules/ uses the best practices.

**Comment by haibeey** [comment]
i wrote a rough prototype to preserve comments by augmenting the goyacc grammar rules a little. commit here [https://github.com/haibeey/prometheus/commit/885566786ac20bfd400b2e7db470c92545919690](https://github.com/haibeey/prometheus/commit/885566786ac20bfd400b2e7db470c92545919690)

[Rules file ](https://github.com/haibeey/haibeey.github.io/blob/master/image/rules.png)
[recording rule in web ui](https://github.com/haibeey/haibeey.github.io/blob/master/image/preservecomment.png)

**Comment by brian-brazil** [comment]
That doesn't look right, the operators aren't on their own line

**Comment by haibeey** [comment]
oh yes. that commit doesn't do formmating. it is just to preserve comments in promql expr after evaluation for printing. 
previous versions ignores all comments after evaluation.

**Comment by slrtbtfs** [comment]
> i wrote a rough prototype to preserve comments by augmenting the goyacc grammar rules a little. commit here [haibeey@8855667]

Putting the comments in the AST seems like a reasonable idea.

I've left some comments about the implementation there.

**Comment by sylr** [comment]
> https://prometheus.io/docs/practices/rules/ uses the best practices.

Please don't enfoce:
```
sum without (instance)(instance_path:requests:rate5m{job="myjob"})
```
over
```
sum(instance_path:requests:rate5m{job="myjob"}) without (instance)
```

**Comment by brian-brazil** [comment]
Having grouping modifiers first is the best practice, and is already how we print it - it's very difficult to read non-trivial expressions otherwise.

**Comment by juliusv** [comment]
Yeah, I had strong opinions about this initially. At first Prometheus only supported `sum(x) by(y)` because it read more like English to me and thus I much preferred it, but for any non-trivial `x` it becomes really hard to see what dimensions you are aggregating over / keeping. I wish I had just made it `sum by(y) (x)` from the beginning and made that the only legal variant.

I still don't like at all the exact formatting `sum by (y)(x)` and much prefer `sum by(y) (x)` (makes much more sense to me), not sure why we chose to go for the latter one in the docs.

**Comment by sylr** [comment]
I think everyone is entitled to its own opinion about this, I personally find the "best practice" to be hideous, but, as all syntaxes are legal, please do not enforce one over the others.

**Comment by beorn7** [comment]
The whole point of a `promfmt` tool is to, if not enforce, at least "endorse" a certain "canonical" formatting, very much like `gofmt`, with a whole string of arguments behind it.

**Comment by juliusv** [comment]
Yeah, `gofmt`'s formatting is nobody's favorite, but at least it creates one consistent style rather than many different ones, which is a value in itself.

**Comment by codesome** [comment]
As a side note, this is actively been sought after for the GSoC. @haibeey I see that you are also interested in GSoC. While I appreciate the upfront work that you have put now, I would appreciate some open discussions and design sharing, in the GSoC proposal or otherwise, for easier coordination of projects in GSoC :)

**Comment by Harkishen-Singh** [comment]
Comment preserving is already supported very well in the new go-yamlv3 library. The implementation is so flawless that there is no loss of comments. Hence, we can keep preserving mechanism independent of prometheus.

**Comment by haibeey** [comment]
alright. i guess it is time share my proposal draft [GSoC proposal](https://docs.google.com/document/d/1H3wVDB-BGqM7OFm4BZBi_JFMS1ukCCVXP9WSq5odUEQ/edit?usp=sharing) .

Reviews be appreciated before the final draft. Thanks

**Comment by juliusv** [comment]
@Harkishen-Singh There are two levels of comments: YAML comments and comments within a PromQL expression. https://github.com/prometheus/prometheus/issues/21#issuecomment-604213849 was talking about the PromQL ones (but YAML ones would be good to preserve as well).

**Comment by haibeey** [comment]
> Comment preserving is already supported very well in the new go-yamlv3 library. The implementation is so flawless that there is no loss of comments. Hence, we can keep preserving mechanism independent from the Prometheus code without any worries.

During evaluation of a PromQL expression is when the  comments are removed.
That said i think the formatting is not only to be done in rules files but in some places too like the web ui. is this true @juliusv ?
So as pointer out by @slrtbtfs most of the formatting  and preserving would be done in the Printer module.

**Comment by brian-brazil** [comment]
In relation to YAML comments, there's problems with v3 that's causing issues for Cortex so we should avoid further rollout of that until that's all resolved.

> That said i think the formatting is not only to be done in rules files but in some places too like the web ui. is this true 

The web ui is the only place it's done.

**Comment by krajorama** [comment]
Hello from the bug scrub!

It seems like this is not happening. Nobody took the task. We're closing and if someone actually needs this, they should open a new issue to put it on top of the issues list for people.

**Comment by krajorama** [comment]
If this gets reopened/renewed, you can build on https://github.com/prometheus/prometheus/pull/11411
